src/Primitives/src/ChangeToken.cs

```diff
@@ -67,26 +50,91 @@ public static IDisposable OnChange<TState>(Func<IChangeToken> changeTokenProduce
+            return new ChangeTokenRegistration<TState>(changeTokenProducer, changeTokenConsumer, state);
+        }
+
+        private class ChangeTokenRegistration<TState> : IDisposable
+        {
+            private readonly Func<IChangeToken> _changeTokenProducer;
+            private readonly Action<TState> _changeTokenConsumer;
+            private readonly TState _state;
+            private object _disposable;
+
+            private static readonly object _disposedSentinel = new object();
+
+            public ChangeTokenRegistration(Func<IChangeToken> changeTokenProducer, Action<TState> changeTokenConsumer, TState state)
+            {
+                _changeTokenProducer = changeTokenProducer;
+                _changeTokenConsumer = changeTokenConsumer;
+                _state = state;
+
+                var token = changeTokenProducer();
+
+                RegisterChangeTokenCallback(token);
+            }
+
+            private void OnChangeTokenFired()
             {
                 // The order here is important. We need to take the token and then apply our changes BEFORE
                 // registering. This prevents us from possible having two change updates to process concurrently.
                 //
                 // If the token changes after we take the token, then we'll process the update immediately upon
                 // registering the callback.
-                var t = changeTokenProducer();
+                var token = _changeTokenProducer();
+
                 try
                 {
-                    changeTokenConsumer((TState)s);
+                    _changeTokenConsumer(_state);
                 }
-                finally // We always want to ensure the callback is registered
+                finally
                 {
-                    t.RegisterChangeCallback(callback, s);
+                    // We always want to ensure the callback is registered
+                    RegisterChangeTokenCallback(token);
                 }
-            };
+            }
+
+            private void RegisterChangeTokenCallback(IChangeToken token)
+            {
+                var registraton = token.RegisterChangeCallback(s => ((ChangeTokenRegistration<TState>)s).OnChangeTokenFired(), this);
+
+                SetDisposable(registraton);
+            }
 
-            return changeTokenProducer().RegisterChangeCallback(callback, state);
+            private void SetDisposable(IDisposable disposable)
+            {
+                // We don't want to transition from _disposedSentinel => anything since it's terminal
+                // but we want to allow going from previously assigned disposable, to another
+                // disposable. The assumption here is that we don't have to dispose the registration if an IChangeToken
+                // fires.
+                var current = Volatile.Read(ref _disposable);
+
+                var previous = Interlocked.CompareExchange(ref _disposable, disposable, current);
```

Comment:
Not sure what `previous` and `current` are achieving here?

The `CompareExchange` would prevent something between the two; but doesn't add anything over the single statement:
```csharp
var previous = Interlocked.Exchange(ref _disposable, disposable);
```
So I assume its trying to do something else?

Reply:
I want to assign _disposable to disposable only if current is not _disposedSentinel

Reply:
e.g. is should the whole method be:
```csharp
var previous = Interlocked.CompareExchange(ref _disposable, disposable, null);

if (previous == _disposedSentinel)
{
    // The subscription was disposed so we dispose immediately and return
    disposable.Dispose();
}
else if (previous != null)
{
    Debug.Fail("Somebody else set the _disposable");
}
```

Reply:
That’s not what I want because SetDisposable can be called multiple times without calling Dispose which means you can transition from non null disposable to disposable. That’s why I used current instead of null.